sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/FileScanBuilder.scala

```diff
@@ -26,34 +26,40 @@ abstract class FileScanBuilder(
     fileIndex: PartitioningAwareFileIndex,
     dataSchema: StructType) extends ScanBuilder with SupportsPushDownRequiredColumns {
   private val partitionSchema = fileIndex.partitionSchema
+  private val partitionNameSet = toNameSet(partitionSchema)
   private val isCaseSensitive = sparkSession.sessionState.conf.caseSensitiveAnalysis
   protected var requiredSchema = StructType(dataSchema.fields ++ partitionSchema.fields)
 
   override def pruneColumns(requiredSchema: StructType): Unit = {
-    this.requiredSchema = requiredSchema
+    // the default implementation prunes only top-level columns
+    // file formats that support nested column pruning must override this
+    val fields = dataSchema.fields ++ partitionSchema.fields
```

Comment:
If partition and data columns overlap, `requiredSchema` might contain duplicates as we match by name. However, `readDataSchema` and `readPartitionSchema` will still report correct structs.

As an alternative to the current approach, we can keep the previous implementation of `pruneColumns` and come up with a new `ScanBuilder` that extends `FileScanBuilder` and overrides `readDataSchema` to iterate through `requiredSchema` instead of `dataSchema`.

Reply:
There are test cases for "partition and data columns overlap" in OrcPartitionDiscoveryTest and `ParquetPartitionDiscoverySuite`. 
Since we are touching the code here, it would be great if there are such test cases for JSON/CSV data source

Reply:
What about extending `FileBasedDataSourceSuite` to cover csv as well? It seems there is a test for orc/parquet/json already.

Reply:
+1 for @gengliangwang 's advice.

Reply:
> What about extending FileBasedDataSourceSuite to cover csv as well? It seems there is a test for orc/parquet/json already.

Yes. And it would be great if there is a new test case with nested column pruning.

Reply:
Sounds good, will add a new one!

Reply:
@gengliangwang @dongjoon-hyun what about the overall approach here? Shall we keep it as is and let `readDataSchema` and `readPartitionSchema` deduplicate columns or shall we modify `readDataSchema` as suggested in my first comment?

Reply:
I prefer the latter one(the alternative approach). It is easier for understanding.

Reply:
Are we going to add another `ScanBuilder` implementation like the following?
> As an alternative to the current approach, we can keep the previous implementation of pruneColumns and come up with a new ScanBuilder that extends FileScanBuilder and overrides readDataSchema to iterate through requiredSchema instead of dataSchema

Actually, that kind of extensibility is one of the original goals of DSv2 approach and I agree this PR's goal. One concern is that it sounds like we are adding another place handling column resolutions. From the beginning of Parquet DSv2, it was tricky to track because the column resolution (due to the mismatch between physical file schema and required/read schema) happens two places (here and filters). After this PR, are we going to have three places?

Before going further, shall we ping, @rdblue and @cloud-fan , too?

Reply:
Well, we don't necessarily need to extend `FileScanBuilder` with another builder.

We can consider smth as below (I reverted most of the changes in `FileScanBuilder`):

```
abstract class FileScanBuilder(
    sparkSession: SparkSession,
    fileIndex: PartitioningAwareFileIndex,
    dataSchema: StructType) extends ScanBuilder with SupportsPushDownRequiredColumns {
  private val partitionSchema = fileIndex.partitionSchema
  private val isCaseSensitive = sparkSession.sessionState.conf.caseSensitiveAnalysis
  protected var requiredSchema = StructType(dataSchema.fields ++ partitionSchema.fields)

  protected def supportsNestedSchemaPruning: Boolean = false

  override def pruneColumns(requiredSchema: StructType): Unit = {
    this.requiredSchema = requiredSchema
  }

  protected def readDataSchema(): StructType = {
    val requiredNameSet = createRequiredNameSet()
    val schema = if (supportsNestedSchemaPruning) requiredSchema else dataSchema
    val fields = schema.fields.filter { field =>
      val colName = PartitioningUtils.getColName(field, isCaseSensitive)
      requiredNameSet.contains(colName) && !partitionNameSet.contains(colName)
    }
    StructType(fields)
  }

  protected def readPartitionSchema(): StructType = {
    val requiredNameSet = createRequiredNameSet()
    val fields = partitionSchema.fields.filter { field =>
      val colName = PartitioningUtils.getColName(field, isCaseSensitive)
      requiredNameSet.contains(colName)
    }
    StructType(fields)
  }

  private def createRequiredNameSet(): Set[String] =
    requiredSchema.fields.map(PartitioningUtils.getColName(_, isCaseSensitive)).toSet

  private val partitionNameSet: Set[String] =
    partitionSchema.fields.map(PartitioningUtils.getColName(_, isCaseSensitive)).toSet
}

```

Reply:
I've updated the PR to reflect what I meant. The old version is still accessible in the commit history.